### PR TITLE
Enhance test for database upgrade process

### PR DESCRIPTION
**Pull request by korelstar**
As proposed in https://github.com/k9mail/k-9/pull/2158#issuecomment-277963119, I've extended the test for the database upgrade process, which was introduced in #2158. There, the test was restricted on the existence of tables, indexes and triggers. Now, the test also checks the structure of these entities using the `sql` column in `sqlite_master`.

Fortunately, I've only found one issue with this new test, and that is minor. It concerns the column `notify_class` of table `folders`. Using the upgrade process, this column is created with `default 'INHERITED'`. However, when the database is created from scratch, the default is missing.

Therefore, I've added the default value to `dbCreateDatabaseFromScratch` to fix the test.

However, this approach show also the lack of current tests: They test only if upgrading from a *virtual* v29 database comes to the same result like creating a new one from scratch (see #2158). But, additionally, a respective test would be nice for **every** release (not only v29)! Because you have to note, that my solution for the `notify_class` column only applies to future new installations. People with a former new installation of a release with database v50 and above will still not have the correct default value. Nevertheless, I've chosen that approach, because the default should not be that important and we would have to re-create the whole table in order to change it. But the message is, that there may be much more problems of this type. They can only be detected with a test of the upgrade-paths of every release! However, that's not that easy, but I have some ideas how to realize this. I will write them down in a new issue in the next days. But for now, this pull request is good enhancement :smile:

### CODE REVIEW

**Comment by philipwhiuk** [comment]
I don't think we even did upgrades before v29 - we may have simply required re-creation - I would not in general spend too long worrying about them.

What I think we should do first is stop destroying databases when the upgrade fails and instead persuade users to raise bugs with the error. Then we can fix them and they can upgrade, saving their data.

**Comment by korelstar** [comment]
Sorry, I was not talking about versions before v29. I was talking about the releases between v29 and now!

However, the other part sounds good, but I'm too new to assess this ;-)

**Comment by philipwhiuk** [comment]
I've just had a look at the code and we don't need to worry about the change in default for `notify_class`. We always provide one when creating a folder anyway.

File: k9mail/src/test/java/com/fsck/k9/mailstore/StoreSchemaDefinitionTest.java [hunk 1]
```diff
@@ -295,11 +297,12 @@ private void assertDatabaseIndexesEquals(SQLiteDatabase expected, SQLiteDatabase
 
     private List<String> objectsInDatabase(SQLiteDatabase db, String type) {
         List<String> tables = new ArrayList<>();
-        Cursor cursor = db.rawQuery("SELECT name FROM sqlite_master WHERE type = ?", new String[] { type });
+        Cursor cursor = db.rawQuery("SELECT sql FROM sqlite_master WHERE type = ? AND sql IS NOT NULL", new String[] { type });
```
Comment (korelstar) [inline k9mail/src/test/java/com/fsck/k9/mailstore/StoreSchemaDefinitionTest.java]: Good point. I've found this for the index `sqlite_autoindex_messages_fulltext_segdir_1` which is automatically created internally by SQLite when creating the FTS4 table `messages_fulltext`. Therefore, we can safely ignore such a type of index.

File: k9mail/src/test/java/com/fsck/k9/mailstore/StoreSchemaDefinitionTest.java [hunk 2]
```diff
@@ -295,11 +297,12 @@ private void assertDatabaseIndexesEquals(SQLiteDatabase expected, SQLiteDatabase
...
         try {
             if (cursor.moveToFirst()) {
                 while (!cursor.isAfterLast()) {
-                    tables.add(cursor.getString(cursor.getColumnIndex("name")));
+                    String sql = cursor.getString(cursor.getColumnIndex("sql"));
+                    tables.add("table".equals(type) ? sortTableColumns(sql) : sql);
```
Comment (korelstar) [inline k9mail/src/test/java/com/fsck/k9/mailstore/StoreSchemaDefinitionTest.java]: Should I change the value for the parameter? We could use an `Enum` here.

Comment (cketti) [inline k9mail/src/test/java/com/fsck/k9/mailstore/StoreSchemaDefinitionTest.java]: Let's not overengineer. Renaming `tables` to e.g. `databaseObjects` will align with the method name. Additionally, you could add a separate method for reading tables from the database. That will lead to some code duplication but reduces complexity inside the individual methods.

Comment (korelstar) [inline k9mail/src/test/java/com/fsck/k9/mailstore/StoreSchemaDefinitionTest.java]: I don't like code duplication in this case. But the other remarks are integrated, now. Please review :-)

**Comment by cketti** [comment]
Thanks!

**Pull request merged**